app/src/org/commcare/services/CommCareFirebaseMessagingService.java

```diff
@@ -28,81 +29,95 @@
+     * to the action it carries
+     */
     @Override
     public void onMessageReceived(RemoteMessage remoteMessage) {
-        Logger.log(LogTypes.TYPE_FCM, "Message received: " + remoteMessage.getMessageId());
+        Logger.log(LogTypes.TYPE_FCM, "CommCareFirebaseMessagingService Message received: " + remoteMessage.getData());
         Map<String, String> payloadData = remoteMessage.getData();
-        RemoteMessage.Notification payloadNotification = remoteMessage.getNotification();
-
-        if (payloadNotification != null) {
-            showNotification(payloadNotification);
-        }
+//      RemoteMessage.Notification payloadNotification = remoteMessage.getNotification();
+//      if (payloadNotification != null) {
+//          showNotification(payloadNotification, payloadData);
+//      }
 
         // Check if the message contains a data object, there is no further action if not
-        if (payloadData.size() == 0){
+        if (payloadData.isEmpty()) {
             return;
         }
 
+        showNotification(payloadData);
+
         FCMMessageData fcmMessageData = new FCMMessageData(payloadData);
 
-        switch(fcmMessageData.getAction()){
+        switch (fcmMessageData.getAction()) {
             case SYNC -> dataSyncer.syncData(fcmMessageData);
-            default ->
-                    Logger.log(LogTypes.TYPE_FCM, "Invalid FCM action");
+            default -> Logger.log(LogTypes.TYPE_FCM, "Invalid FCM action");
         }
     }
 
     @Override
     public void onNewToken(String token) {
         // TODO: Remove the token from the log
-        Logger.log(LogTypes.TYPE_FCM, "New registration token was generated: "+token);
+        Logger.log(LogTypes.TYPE_FCM, "New registration token was generated: " + token);
         FirebaseMessagingUtil.updateFCMToken(token);
     }
 
 
-   /**
-    * This method purpose is to show notifications to the user when the app is in the foreground.
-    * When the app is in the background, FCM is responsible for notifying the user
-    *
-    */
-    private void showNotification(RemoteMessage.Notification notification) {
-        String notificationTitle = notification.getTitle();
-        String notificationText = notification.getBody();
+    /**
+     * This method purpose is to show notifications to the user when the app is in the foreground.
+     * When the app is in the background, FCM is responsible for notifying the user
+     */
+    private void showNotification(Map<String, String> payloadData) {
+        String notificationTitle = payloadData.get("title");
+        String notificationText = payloadData.get("body");
         NotificationManager mNM = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
 
-        Intent i = new Intent(this, DispatchActivity.class);
-        i.setAction(Intent.ACTION_MAIN);
-        i.addCategory(Intent.CATEGORY_LAUNCHER);
+        Intent intent;
+
+        if (payloadData.containsKey("action")) {
```

Comment:
Will this code behave correctly for the existing SYNC action that was sent in existing CommCare, i.e. not related to Connect? Need to make sure we aren't passing the SYNC action on to Connect instead of using the default path

Reply:
It should work because I didn't change the existing sync logic. I just added a condition: if the payload contains the "action" keyword, it redirects to **ConnectActivity**. Otherwise, it redirects to **DispatchActivity**, as previously set.

Reply:
I'm still worried that the SYNC command also uses the 'action' field. Have a look at the constructor for FCMMessageData, which is called just after this function (see lines 62-64). It retrieves this.action from the 'action' field in the payloadData. So wouldn't we then incorrectly route this notification to the ConnectActivity?